py/server/tests/test_partitioned_table_service.py

```diff
@@ -0,0 +1,248 @@
+        if table_key.key == "test":
+            return self.partitions[table_location_key].select([col]).slice(offset, max_rows)
+        else:
+            return pa.table([])
+
+    def _th_new_partitions(self, table_key: TableKey, exec_ctx: ExecutionContext, callback: Callable[[PartitionedTableLocationKey, Optional[pa.Table]], None]) -> None:
+        if table_key.key != "test":
+            return
+
+        while not self.sub_new_partition_cancelled:
+            try:
+                with exec_ctx:
+                    pa_table = next(self.gen_pa_table)
+            except StopIteration:
+                break
+
+            ticker = str(pa_table.column("Ticker")[0])
+            partition_key = PartitionedTableLocationKey(f"{ticker}/NYSE")
+            self.partitions[partition_key] = pa_table
+
+            expr = ((pc.field("Ticker") == f"{ticker}") & (pc.field("Exchange") == "NYSE"))
+            callback(partition_key, pa_table.filter(expr).select(["Ticker", "Exchange"]).slice(0, 1))
+            time.sleep(0.1)
+
+    def subscribe_to_new_partitions(self, table_key: TableKey, callback) -> Callable[[], None]:
+        if table_key.key != "test":
+            return lambda: None
+
+        exec_ctx = get_exec_ctx()
+        th = threading.Thread(target=self._th_new_partitions, args=(table_key, exec_ctx, callback))
+        th.start()
+
+        def _cancellation_callback():
+            self.sub_new_partition_cancelled += 1
+
+        return _cancellation_callback
+
+
+    def _th_partition_size_changes(self, table_key: TableKey, table_location_key: PartitionedTableLocationKey, callback: Callable[[int], None]) -> None:
+        if table_key.key != "test":
+            return
+
+        if table_location_key not in self.partitions_size_subscriptions:
+            return
+
+        while self.partitions_size_subscriptions[table_location_key]:
+            pa_table = self.partitions[table_location_key]
+            rbs = pa_table.to_batches()
+            rbs.append(pa_table.to_batches()[0])
+            new_pa_table = pa.Table.from_batches(rbs)
+            self.partitions[table_location_key] = new_pa_table
+            callback(new_pa_table.num_rows)
+            time.sleep(0.1)
+
+
+    def subscribe_to_partition_size_changes(self, table_key: TableKey,
+                                            table_location_key: PartitionedTableLocationKey,
+                                            callback: Callable[[int], None]) -> Callable[[], None]:
+        if table_key.key != "test":
+            return lambda: None
+
+        if table_location_key not in self.partitions:
+            return lambda: None
+
+        self.partitions_size_subscriptions[table_location_key] = True
+        th = threading.Thread(target=self._th_partition_size_changes, args=(table_key, table_location_key, callback))
+        th.start()
+
+        def _cancellation_callback():
+            self.partitions_size_subscriptions[table_location_key] = False
```

Comment:
Is this thread safe in 3.13t?  The value is being accessed from two different threads without protection.

Reply:
Yes, the engine should only call the cancellation callback only once for each table location key